### ISSUE TITLE
Pipelining decisions

### ISSUE THREAD

**Issue by apasarkar** [apasarkar/masknmf-toolbox#195]
The backend API is now clean/modular. But on its own, it still requires the end user to chain together a whole analysis pipeline. I'm attaching below an example workflow used to analyze some standard 2p calcium imaging data. 

```
data = masknmf.TiffArray("datapath.tif", memmap=True)

frame_batch_size = 300 #Reduce this if your GPU doesn't have enough memory

#Max shifts is the maximum allowed rigid shift. Main parameter to modify here.
rigid_strategy = masknmf.RigidMotionCorrector(max_shifts = [30, 30], batch_size=frame_batch_size)
rigid_strategy.compute_template(data)
                      
full_moco_arr = masknmf.motion_correction.RegistrationArray(data,
                                                           strategy = rigid_strategy)

full_moco_arr.export("moco.hdf5")
shifts = full_moco_arr.shifts
pixel_weighting = construct_moco_template(full_moco_arr.shifts, full_moco_arr.shape[1:])


## Now run compression:

block_sizes = [20, 20]
max_components = 20
max_consecutive_failures = 1
temporal_avg_factor = 1
spatial_avg_factor = 1
frame_batch_size = 300
pixel_weighting = pixel_weighting
noise_variance_quantile = 0.3
num_epochs = 10
compress_strat = masknmf.CompressDenoiseStrategy(full_moco_arr[:],
                                                 block_sizes=block_sizes,
                                                 max_components=max_components,
                                                 max_consecutive_failures=max_consecutive_failures,
                                                 temporal_avg_factor=temporal_avg_factor,
                                                 spatial_avg_factor=spatial_avg_factor,
                                                 frame_batch_size=frame_batch_size,
                                                 pixel_weighting=pixel_weighting,
                                                 noise_variance_quantile=noise_variance_quantile,
                                                 num_epochs = num_epochs
                                                 )

pmd_denoised = compress_strat.compress()


##Now run multiple passes of demixing

#Step 1: Get a high-pass filtered version of the data, demix that first 

device='cuda'
frame_batch_size = 200
# Generate a spatially filtered version of the PMD data
spatial_filt_pmd = masknmf.demixing.filters.spatial_filter_pmd(pmd_denoise,
                                                   batch_size=frame_batch_size,
                                                   filter_sigma=2,
                                                   device = device)

display("Processing spatially-highpass filtered PMD array")
device = 'cuda'
highpass_pmd_demixer = masknmf.demixing.signal_demixer.SignalDemixer(spatial_filt_pmd,
                                                                     device=device,
                                                                     frame_batch_size=frame_batch_size)

init_kwargs = {
    'mad_correlation_threshold':0.8,

    #Mostly stable
    'mad_threshold':1,
    'residual_threshold': 0.3,
    'patch_size':(40, 40),
}

highpass_pmd_demixer.initialize_signals(**init_kwargs, is_custom = False)
if highpass_pmd_demixer.results is not None:
    display(f"Identified {highpass_pmd_demixer.results.a.shape[1]} candidate neural signals at initialization step.")

# Run NMF
## Demixing State
num_iters = 25
## Now run demixing...
localnmf_params = {
    'maxiter':num_iters,
    'support_threshold':np.linspace(0.95, 0.8, num_iters).tolist(),
    'deletion_threshold':0.2,
    'ring_model_start_pt':num_iters+1, #Allow one support update to occur
    'ring_radius':10,
    'background_downsampling_factor':30,
    'merge_threshold':0.6,
    'merge_overlap_threshold':0.3,
    'update_frequency':4,
    'c_nonneg':True,
    'denoise':False,
    'plot_en': False
}

with torch.no_grad():
    highpass_pmd_demixer.demix(**localnmf_params)
display(f"After demixing the high-pas data, there were {highpass_pmd_demixer.results.a.shape[1]} signals identified")

## Find an affine transform of these signals to match the unfiltered PMD data and demix that using the above footprints


a_init = highpass_pmd_demixer.results.ac_array.export_a()
c_init = highpass_pmd_demixer.results.ac_array.export_c()
# a_init = hp_demix_results.ac_array.export_a()
# c_init = hp_demix_results.ac_array.export_c()

frame_batch_size = 200
device = 'cuda'
unfiltered_pmd_demixer = masknmf.demixing.signal_demixer.SignalDemixer(
                                                pmd_denoise,
                                                device=device,
                                                frame_batch_size=frame_batch_size)

unfiltered_pmd_demixer.initialize_signals(is_custom=True,
                                          spatial_footprints=a_init,
                                          temporal_footprints=c_init,
                                          c_nonneg = True)

## Demixing State
num_iters = 25
## Now run demixing...
localnmf_params = {
    'maxiter':num_iters,
    'support_threshold':np.linspace(0.95, 0.5, num_iters).tolist(),
    'deletion_threshold':0.2,
    'ring_model_start_pt':0,
    'ring_radius': 15,
    'background_downsampling_factor': 20,
    'merge_threshold':0.8,
    'merge_overlap_threshold':0.8,
    'update_frequency':4,
    'c_nonneg':True,
    'denoise':False,
    'plot_en': False
}

with torch.no_grad():
    unfiltered_pmd_demixer.demix(**localnmf_params)
display(f"after this step {unfiltered_pmd_demixer.results.a.shape[1]} signals identified")


## Run another pass of demixing to find remaining signals. Start with initialization

init_kwargs = {
    'mad_correlation_threshold':0.7,
    'mad_threshold':1,
    'residual_threshold': 0.3,
    'patch_size':(40, 40),
}

unfiltered_pmd_demixer.initialize_signals(**init_kwargs, carry_background=True, is_custom = False)
if unfiltered_pmd_demixer.results is not None:
    display(f"Identified {unfiltered_pmd_demixer.results.a.shape[1]} candidate neural signals at initialization step.")

##Run the NMF 

num_iters = 40
## Now run demixing...
localnmf_params = {
    'maxiter':num_iters,
    'support_threshold':np.linspace(0.95, 0.4, num_iters).tolist(),
    'deletion_threshold':0.2,
    'ring_model_start_pt':0,
    'ring_radius':5,
    'background_downsampling_factor':10,
    'merge_threshold':0.8,
    'merge_overlap_threshold':0.8,
    'update_frequency':4,
    'c_nonneg':True,
    'denoise':False,
    'plot_en': False
}

with torch.no_grad():
    unfiltered_pmd_demixer.demix(**localnmf_params)
display(f"after this step {unfiltered_pmd_demixer.results.a.shape[1]} signals identified")
torch.cuda.empty_cache()



```

What's annoying about the above? 
- Lots of params sitting in dictionaries. dataclasses are nice for this. Users might have specific things they need to do (some might want to mask out certain pixels, etc. they can adjust the default data class appropriately). 
- Have to keep track of way too much information (which params have changed, what is the state of the signal demixer, etc.)

With configs, the above can be replaced by: 

```
def standard_twophoton_calcium_pipeline(data: Union[np.ndarray, LazyFrameLoader, ArrayLike],
                                        motion_correct_config: RigidMotionCorrectionConfig | PiecewiseRigidMotionCorrectionConfig | Literal["skip"] | None = None,
                                        compress_config: CompressConfig | CompressDenoiseConfig | None = None,
                                        spatial_highpass_config: SpatialHighpassConfig | None = None,
                                        filtered_demixing_config: MultipassDemixingConfig | None = None,
                                        unfiltered_demixing_config: MultipassDemixingConfig | None = None,
                                        outpath_motion_correction: Optional[str] = "motion_correction.hdf5",
                                        outpath_compression: Optional[str] = "compression.hdf5",
                                        outpath_demixing: Optional[str] = "demixing_results.hdf5",
                                        load_into_ram: bool = True,
                                        frame_batch_size: int= 300,
                                        device: Literal["auto", "cuda", "cpu"] = "auto") -> masknmf.DemixingResults:

```

This entrypoint can be run in "hello world" fashion (just passing in the only required field - "data"). Or it can be fully customized to produce something as expressive as the above code. 

@kushalkolar thoughts?
